qpid-jms-client/src/test/java/org/apache/qpid/jms/test/testpeer/TestAmqpPeer.java

```diff
@@ -495,6 +512,169 @@ public void run()
+                                }
+                            });
+                        } catch (PrivilegedActionException e) {
+                            e.printStackTrace();
+                        }
+                        LOGGER.info("Complete:" + saslServer.isComplete());
+
+                        if (token != null) {
+                            // fling it back in on complete
+                            challengeFrame.setChallenge(new Binary(token));
+                        }
+                        return true;
+                    }
+                }).onCompletion(new AmqpPeerRunnable() {
+                    @Override
+                    public void run() {
+                        TestAmqpPeer.this.sendFrame(
+                                FrameType.SASL, 0,
+                                challengeFrame,
+                                null,
+                                false, 0);
+                    }
+                });
+
+        AtomicBoolean response = new AtomicBoolean(false);
+        SaslResponseMatcher challengeMatcher = new SaslResponseMatcher().withResponse(new BaseMatcher<Binary>() {
+
+            @Override
+            public void describeTo(Description description) {}
+
+            @Override
+            public boolean matches(Object o) {
+                final Binary binary = (Binary) o;
+                if (!saslServer.isComplete()) {
+                    // validate via sasl
+                    byte[] token = null;
+                    try {
+                        token = Subject.doAs(serverSubject, new PrivilegedExceptionAction<byte[]>() {
+                            @Override
+                            public byte[] run() throws Exception {
+                                LOGGER.info("Evaluate challenge response.. size:" + binary.getLength());
+                                return saslServer.evaluateResponse(binary.getArray());
+                            }
+                        });
+                    } catch (PrivilegedActionException e) {
+                        e.printStackTrace();
+                    }
+                    LOGGER.info("Complete:" + saslServer.isComplete());
+                    if (token != null) {
+                        // fling it back
+                        challengeFrame.setChallenge(new Binary(token));
+                        response.set(true);
+                    }
+                    return true;
+                }
+
+                return false;
+            }
+        }).onCompletion(new AmqpPeerRunnable() {
+            @Override
+            public void run() {
+                if (response.get()) {
+                    TestAmqpPeer.this.sendFrame(
+                            FrameType.SASL, 0,
+                            challengeFrame,
+                            null,
+                            false, 0);
+                }
+
+                if (saslServer.isComplete()) {
```

Comment:
if it goes on its own it suggests the other bit can be removed or theres an 'else' missing, which doesnt make sense either as there should likely always be an outcome. I also missed earlier that it is sending a successfull auth outcome..the server being complete doesnt necessarily mean it succeeded, and the previous bit allows for it to fail, so it still seems something is off here to me.